segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ThrottlerCalculator.java

```diff
@@ -0,0 +1,217 @@
+     */
+    @VisibleForTesting
+    static final int MAX_DELAY_MILLIS = 10000;
+    /**
+     * Amount of time (millis) to increase throttling by for each percentage point increase in the cache utilization (above 100%).
+     */
+    @VisibleForTesting
+    static final double THROTTLING_MILLIS_PER_PERCENT_OVER_LIMIT = 100;
+
+    @Singular
+    private final List<Throttler> throttlers;
+
+    //endregion
+
+    //region Methods
+
+    /**
+     * Determines whether there is an immediate need to introduce a throttling delay.
+     *
+     * @return True if throttling is required, false otherwise.
+     */
+    boolean isThrottlingRequired() {
+        for (Throttler t : this.throttlers) {
+            if (t.isThrottlingRequired()) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Calculates the amount of time needed to delay based on the configured throttlers. The computed result is not additive,
+     * as there is no benefit to adding delays from various Throttle Calculators together. For example, a cache throttling
+     * delay will have increased batching as a side effect, so there's no need to include the batching one as well.
+     *
+     * @return A DelayResult representing the computed delay. This delay is the maximum delay as calculated by the internal
+     * throttlers.
+     */
+    DelayResult getThrottlingDelay() {
+        // These delays are not additive. There's no benefit to adding a batching delay on top of a throttling delay, since
+        // a throttling delay will have increased batching as a side effect.
+        int maxDelay = 0;
+        boolean maximum = false;
+        for (Throttler t : this.throttlers) {
+            DelayResult r = t.getDelay();
+            maxDelay = Math.max(maxDelay, r.durationMillis);
+            maximum |= r.maximum;
+        }
+
+        return new DelayResult(maxDelay, maximum);
+    }
+
+    //endregion
+
+    //region Throttlers
+
+    /**
+     * Defines the structure of a Throttle Calculator.
+     */
+    static abstract class Throttler {
+        /**
+         * Determines whether throttling based on this calculator is absolutely required at this moment.
+         */
+        abstract boolean isThrottlingRequired();
+
+        /**
+         * Calculates a throttling delay based on information available at the moment.
+         */
+        abstract DelayResult getDelay();
+    }
+
+    /**
+     * Calculates the amount of time to wait before processing more operations from the queue in order to relieve pressure
+     * on the cache. This is based on ReadIndex statistics, mainly cache utilization ratio.
+     */
+    @RequiredArgsConstructor
+    private static class CacheThrottler extends Throttler {
+        private final Supplier<Double> getCacheUtilization;
+
+        @Override
+        boolean isThrottlingRequired() {
+            return this.getCacheUtilization.get() > 1;
```

Comment:
The thinking here is to set the cache capacity at something lower than our actual physical capacity. Suppose we have a 10GB physical limit (memory or SSD) that we dedicate to RocksDB. We need to set the max cache limit to 75% of that. When we reach that limit, the Cache Manager will start evicting evictable data and only keep the data that is needed.

Doing what you said will have unintended side effects. The Cache Manager eviction policy is that it doesn't touch the cache until it reaches it limit, since it wants to make full use of the cache, vs just a portion of it. Using this strategy, we would inadvertently start throttling for no reason, since the cache can be nearing capacity without there being any Tier2 backlog.

 Fixing this would require the Cache Manager to collect and report both evictable and unevictable data, a considerable increase in complexity which I was not willing to introduce for this reason, especially since we have a much more elegant approach available to us.

Reply:
I'm not suggesting that we use only 80% or whatever other value under 100% of the cache, but the goal is to avoid reaching 100% as much as possible because if the cache is fully populated, then it needs to evict before bringing anything new in. A typical strategy is to start throttling earlier so that we still don't have to evict (assuming the eviction process is more costly compared to the throttling) to accommodate new data.

Reply:
> then it needs to evict before bringing anything new in

That's not how this particular cache works. It will not block inserts when the capacity is reached. The Cache Manager runs on a periodic basis (5 seconds) and queries all registered caches (each Segment Index). If the total sum of each of those caches exceeds the limit, it instructs them to evict the oldest data (until the sum is below the limit or there is nothing more it can do). After each such run, it updates the stats that the Throttler Calculator uses to determine the appropriate throttling mechanism.

If you want this cache to check on each insert whether it exceeds the capacity and possibly block that, then we are talking about a much more complex mechanism which will only introduce problems.